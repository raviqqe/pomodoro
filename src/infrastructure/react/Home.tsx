import React from "react";
import { PulseLoader } from "react-spinners";
import styled from "styled-components";
import { SignOut } from "./SignOut";

const Container = styled.div`
  display: flex;
  flex-direction: column;
  height: 100vh;
  width: 85ex;
  max-width: 100%;
  margin: auto;

  > :first-child {
    flex: 1;
  }
`;

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
`;

const SignOutContainer = styled.div`
  position: fixed;
  top: 0.5em;
  right: 0.5em;
`;

export interface IProps {
  signOut: () => Promise<void>;
}

export const Home = ({ signOut }: IProps) => {
  return (
    <Container>
      <LoaderContainer>
        <PulseLoader color="white" />
      </LoaderContainer>
      <SignOutContainer>
        <SignOut signOut={signOut} />
      </SignOutContainer>
    </Container>
  );
};
