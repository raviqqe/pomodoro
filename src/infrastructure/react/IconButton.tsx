import React, { AriaAttributes, ReactNode } from "react";
import styled from "styled-components";

const Button = styled.div`
  text-shadow: 0rem 0.2rem 0.2rem rgba(0, 0, 0, 0.2);
  color: grey;
  cursor: pointer;
  font-size: 1.5em;
  display: flex;
`;

interface IProps extends AriaAttributes {
  children: ReactNode;
  onClick: () => void;
}

export const IconButton = ({ children, onClick, ...restProps }: IProps) => (
  <Button onClick={onClick} {...restProps}>
    {children}
  </Button>
);
