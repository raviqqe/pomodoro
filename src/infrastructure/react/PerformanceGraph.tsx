import { last } from "lodash";
import { DateTime } from "luxon";
import {
  CartesianGrid,
  ResponsiveContainer,
  XAxis,
  YAxis,
  Label,
  Bar,
  BarChart,
} from "recharts";
import styled from "styled-components";
import { IPerformanceGraph } from "../../application/performance-graph";
import { DateSerializer } from "../../domain/date-serializer";
import { white } from "./style/colors";

const Container = styled.div`
  width: 80vw;
  height: 80vh;
`;

const Message = styled.div`
  color: ${white};
  font-size: 1.5em;
`;

export interface IProps {
  performanceGraph: IPerformanceGraph;
}

export const PerformanceGraph = ({
  performanceGraph: { data },
}: IProps): JSX.Element =>
  data.length === 0 ? (
    <Message>No performance graph to show yet!</Message>
  ) : (
    <Container>
      <ResponsiveContainer>
        <BarChart data={data}>
          <CartesianGrid fill="white" stroke="grey" strokeDasharray="3 3" />
          <XAxis
            dataKey="date"
            stroke="grey"
            tickFormatter={(date: string): string => {
              const days: number = DateTime.fromJSDate(
                DateSerializer.deserialize(
                  (last(data) as { date: string }).date
                )
              ).diff(
                DateTime.fromJSDate(DateSerializer.deserialize(date)),
                "days"
              ).days;

              return days === 0 ? "Today" : `${days} days ago`;
            }}
            tickMargin={10}
          />
          <YAxis allowDecimals={false} stroke="grey" tickMargin={5}>
            <Label
              angle={-90}
              position="insideLeft"
              style={{ fill: "grey" }}
              value="Pomodoros"
            />
          </YAxis>
          <Bar dataKey="pomodoros" fill="salmon" />
        </BarChart>
      </ResponsiveContainer>
    </Container>
  );
