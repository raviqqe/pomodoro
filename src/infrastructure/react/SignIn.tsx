import React from "react";
import { TextButton } from "./TextButton";

interface IProps {
  signIn: () => void;
}

export const SignIn = ({ signIn }: IProps) => (
  <TextButton onClick={signIn}>Sign in</TextButton>
);
