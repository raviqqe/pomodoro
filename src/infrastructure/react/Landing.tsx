import { GoMarkGithub } from "react-icons/go";
import React from "react";
import styled from "styled-components";
import { SignIn } from "./SignIn";

const Container = styled.div`
  height: 100vh;
  width: 100vw;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;

  > * {
    margin: 2rem;
  }
`;

const Title = styled.div`
  color: steelblue;
  font-family: "Chelsea Market", sans-serif;
  font-size: 4em;
  font-weight: bold;
  letter-spacing: -0.35ex;
  text-shadow: 0.4em 0.4em 0.4em rgba(0, 0, 0, 0.1);
`;

const GitHubLink = styled.a`
  font-size: 2.5rem;
  margin: 1.5rem;
  color: #222;
  display: block;
  position: fixed;
  bottom: 0em;
  right: 0em;
  line-height: 0ex;
`;

export interface IProps {
  repositoryURL: string;
  signIn: () => void;
}

export const Landing = ({ repositoryURL, signIn }: IProps) => (
  <Container>
    <Title>P🍅m🍅d🍅r🍅</Title>
    <SignIn signIn={signIn} />
    <GitHubLink href={repositoryURL} target="_blank">
      <GoMarkGithub />
    </GitHubLink>
  </Container>
);
