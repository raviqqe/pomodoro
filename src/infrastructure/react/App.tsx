import { Omit } from "lodash";
import { PulseLoader } from "react-spinners";
import { useAsync } from "react-use";
import React, { useState } from "react";
import styled from "styled-components";
import { IInitialState } from "../../application/application-initializer";
import { IProps as ILandingProps, Landing } from "./Landing";
import { Home, IProps as IHomeProps } from "./Home";

const LoaderContainer = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  height: 100vh;
  width: 100vw;
`;

interface IProps extends Omit<IHomeProps, "signOut">, ILandingProps {
  initialize: () => Promise<IInitialState>;
  signIn: () => Promise<boolean>;
  signOut: () => Promise<boolean>;
}

export const App = ({
  initialize,
  repositoryURL,
  signIn,
  signOut,
  ...props
}: IProps) => {
  const [signedIn, setSignedIn] = useState<boolean | null>(null);

  useAsync(async () => {
    const { signedIn } = await initialize();
    setSignedIn(signedIn);
  }, []);

  if (signedIn === null) {
    return (
      <LoaderContainer>
        <PulseLoader color="white" />
      </LoaderContainer>
    );
  } else if (!signedIn) {
    return (
      <Landing
        repositoryURL={repositoryURL}
        signIn={async () => setSignedIn(await signIn())}
      />
    );
  }

  return <Home {...props} signOut={async () => setSignedIn(await signOut())} />;
};
