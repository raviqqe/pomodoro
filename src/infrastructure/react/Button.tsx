import { FC, HTMLAttributes, HTMLProps, ReactNode } from "react";
import styled from "styled-components";
import { boxShadow } from "./style";

interface IProps
  extends HTMLProps<HTMLButtonElement>,
    HTMLAttributes<HTMLButtonElement> {
  secondary?: boolean;
  children?: ReactNode;
}

export const Button: FC<IProps> = styled.button<{ secondary?: boolean }>`
  ${boxShadow};
  background: ${({ secondary }) => (secondary ? "darkgrey" : "salmon")}
  display: flex;
  justify-content: center;
  align-items: center;
  font-family: inherit;
  color: white;
  border: none;
  cursor: pointer;
  flex-shrink: 0;
` as FC<any>;
