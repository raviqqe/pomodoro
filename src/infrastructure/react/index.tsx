import { render } from "react-dom";
import React from "react";
import { ApplicationInitializer } from "../../application/application-initializer";
import { SignInManager } from "../../application/sign-in-manager";
import { SignOutManager } from "../../application/sign-out-manager";
import { PerformanceGraphViewer } from "../../application/performance-graph-viewer";
import { PomodoroTimerStopper } from "../../application/pomodoro-timer-stopper";
import { PomodoroTimerStarter } from "../../application/pomodoro-timer-starter";
import { PomodoroTimerState } from "../../application/pomodoro-timer-state";
import { IRenderer, IPomodoroTimer } from "../renderer";
import { IPerformanceGraph } from "../../application/performance-graph";
import { App, IProps as IAppProps } from "./App";
import { GlobalStyle } from "./style";

interface IPresenter {
  setRenderer(renderer: IRenderer): void;
}

interface IProps
  extends Pick<IAppProps, "performanceGraph" | "signedIn" | "timer"> {}

export class ReactRenderer implements IRenderer {
  private props: IProps = {
    performanceGraph: { data: [] },
    signedIn: null,
    timer: { seconds: 0, state: PomodoroTimerState.Pomodoro, stopped: true },
  };

  constructor(
    private readonly element: HTMLElement,
    presenters: IPresenter[],
    private readonly applicationInitializer: ApplicationInitializer,
    private readonly performanceGraphViewer: PerformanceGraphViewer,
    private readonly pomodoroTimerStarter: PomodoroTimerStarter,
    private readonly pomodoroTimerStopper: PomodoroTimerStopper,
    private readonly signInManager: SignInManager,
    private readonly signOutManager: SignOutManager,
    private readonly repositoryURL: string
  ) {
    for (const presenter of presenters) {
      presenter.setRenderer(this);
    }
  }

  public render(): void {
    this.renderProps({});
  }

  public renderPerformanceGraph(performanceGraph: IPerformanceGraph): void {
    this.renderProps({ performanceGraph });
  }

  public renderSignedIn(signedIn: boolean): void {
    this.renderProps({ signedIn });
  }

  public renderTimer(timer: IPomodoroTimer): void {
    this.renderProps({ timer });
  }

  private renderProps(props: Partial<IProps>): void {
    this.props = { ...this.props, ...props };

    render(
      <>
        <App
          {...this.props}
          initialize={() => this.applicationInitializer.initialize()}
          repositoryURL={this.repositoryURL}
          signIn={() => this.signInManager.signIn()}
          signOut={() => this.signOutManager.signOut()}
          startTimer={() => this.pomodoroTimerStarter.start()}
          stopTimer={() => this.pomodoroTimerStopper.stop()}
          viewGraph={() => this.performanceGraphViewer.viewGraph()}
        />
        <GlobalStyle />
      </>,
      this.element
    );
  }
}
