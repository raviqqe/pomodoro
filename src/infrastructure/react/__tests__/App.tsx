import { fireEvent, render, waitForDomChange } from "@testing-library/react";
import React from "react";
import { App } from "../App";

it("renders before a user signs in", async () => {
  const result = render(
    <App
      initialize={async () => ({ signedIn: false })}
      signIn={async () => true}
      signOut={async () => false}
      repositoryURL="url"
    />
  );

  await waitForDomChange(result);

  expect(result.container).toMatchSnapshot();
});

it("renders after a user signs in", async () => {
  const result = render(
    <App
      initialize={async () => ({ signedIn: true })}
      signIn={async () => true}
      signOut={async () => false}
      repositoryURL="url"
    />
  );

  await waitForDomChange(result);

  expect(result.container).toMatchSnapshot();
});

it("goes to a home screen when a user signs in", async () => {
  const { asFragment, container } = render(
    <App
      initialize={async () => ({ signedIn: false })}
      signIn={async () => true}
      signOut={async () => false}
      repositoryURL="url"
    />
  );

  await waitForDomChange({ container });

  fireEvent.click(container.querySelector("button") as Element);

  await waitForDomChange({ container });

  expect(asFragment()).toMatchSnapshot();
});
