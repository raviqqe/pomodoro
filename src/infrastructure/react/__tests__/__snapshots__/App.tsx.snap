// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders after a user signs in 1`] = `
<div>
  <div
    class="sc-jrAFXE bSzEGg"
  >
    <div
      class="sc-jSgvzq cBMHuY"
    >
      <div
        class="sc-gKseQn fxFXiT"
      >
        🍅
      </div>
      <button
        class="sc-dlfnuX sc-eCstlR enNVDr lkujrQ"
      >
        Start
      </button>
    </div>
    <div
      class="sc-kEjbQP dHBIpC"
    >
      <button
        class="sc-dlfnuX sc-hKgJUU iBalNg gpDZkz"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5a2 2 0 00-2 2v4h2V5h14v14H5v-4H3v4a2 2 0 002 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </button>
      <button
        aria-label="View Graph"
        class="sc-dlfnuX sc-hKgJUU iBalNg gpDZkz"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M176 64h64v384h-64zM80 336h64v112H80zM272 272h64v176h-64zM368 176h64v272h-64z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`renders after a user signs out 1`] = `
<div>
  <div
    class="sc-iqHYmW gdbKBX"
  >
    <div
      class="sc-crrszt jtFBHM"
    >
      P🍅m🍅d🍅r🍅
    </div>
    <button
      class="sc-dlfnuX sc-eCstlR enNVDr lkujrQ"
    >
      Sign in
    </button>
    <a
      class="sc-dQoVA bLThgl"
      href=""
      target="_blank"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"
          fill-rule="evenodd"
        />
      </svg>
    </a>
  </div>
</div>
`;

exports[`renders before a user signs in 1`] = `
<div>
  <div
    class="sc-bqyKOL kJUjYJ"
  >
    <span
      class="css-1xdhyk6"
    >
      <span
        class="css-hqq8du"
      />
      <span
        class="css-1h0qtnz"
      />
      <span
        class="css-1ihj5oh"
      />
    </span>
  </div>
</div>
`;
